File: src/handlers/swift.ts
import axios from 'axios'
import { McpError, ErrorCode } from '@modelcontextprotocol/sdk/types.js'
import {
  PackageVersion,
  SwiftDependency,
  PackageHandler,
  VersionConstraint,
  VersionConstraints
} from '../types/index.js'

export class SwiftHandler implements PackageHandler {
  private async getPackageVersion(
    packageUrl: string,
    currentVersion?: string,
    versionRequirement?: string,
    constraints?: VersionConstraint
  ): Promise<PackageVersion> {
    // Extract package name from URL
    const packageName = packageUrl.split('/').pop()?.replace('.git', '') || packageUrl

    // Check if package should be excluded
    if (constraints?.excludePackage) {
      return {
        name: packageName,
        currentVersion: currentVersion,
        latestVersion: currentVersion || 'unknown',
        registry: 'swift',
        skipped: true,
        skipReason: 'Package excluded from updates'
      }
    }

    try {
      // For GitHub repositories, we can use the GitHub API to get the latest release
      if (packageUrl.includes('github.com')) {
        // Convert GitHub URL to API URL
        // Example: https://github.com/apple/swift-argument-parser -> https://api.github.com/repos/apple/swift-argument-parser/releases
        const apiUrl = packageUrl
          .replace('https://github.com/', 'https://api.github.com/repos/')
          .replace('.git', '') + '/releases'

        const response = await axios.get(apiUrl)

        if (!response.data || response.data.length === 0) {
          throw new Error('No releases found')
        }

        // Find the latest release that's not a pre-release
        let latestRelease = response.data.find((release: any) => !release.prerelease)

        // If no stable release is found, use the latest release
        if (!latestRelease) {
          latestRelease = response.data[0]
        }

        let latestVersion = latestRelease.tag_name

        // Remove 'v' prefix if present
        if (latestVersion.startsWith('v')) {
          latestVersion = latestVersion.substring(1)
        }

        // If major version constraint exists, check if the latest version complies
        if (constraints?.majorVersion !== undefined) {
          const major = parseInt(latestVersion.split('.')[0])
          if (major !== constraints.majorVersion) {
            // Find the latest release within the specified major version
            const constrainedRelease = response.data.find((release: any) => {
              let version = release.tag_name
              if (version.startsWith('v')) {
                version = version.substring(1)
              }
              return parseInt(version.split('.')[0]) === constraints.majorVersion
            })

            if (constrainedRelease) {
              latestVersion = constrainedRelease.tag_name
              if (latestVersion.startsWith('v')) {
                latestVersion = latestVersion.substring(1)
              }
            }
          }
        }

        const result: PackageVersion = {
          name: packageName,
          latestVersion,
          registry: 'swift'
        }

        if (currentVersion) {
          result.currentVersion = currentVersion
        }

        if (constraints?.majorVersion !== undefined) {
          result.skipReason = `Limited to major version ${constraints.majorVersion}`
        }

        return result
      } else {
        // For non-GitHub repositories, we can't easily determine the latest version
        // Return the current version as the latest version
        return {
          name: packageName,
          currentVersion: currentVersion,
          latestVersion: currentVersion || 'unknown',
          registry: 'swift',
          skipped: true,
          skipReason: 'Non-GitHub repository, cannot determine latest version'
        }
      }
    } catch (error) {
      console.error(`Error fetching Swift package ${packageName}:`, error)
      throw new McpError(
        ErrorCode.InternalError,
        `Failed to fetch Swift package ${packageName}`
      )
    }
  }

  async getLatestVersion(args: {
    dependencies: SwiftDependency[],
    constraints?: VersionConstraints
  }) {
    if (!args.dependencies || !Array.isArray(args.dependencies)) {
      throw new McpError(
        ErrorCode.InvalidParams,
        'Invalid dependencies array'
      )
    }

    const results: PackageVersion[] = []
    for (const dependency of args.dependencies) {
      try {
        const result = await this.getPackageVersion(
          dependency.url,
          dependency.version,
          dependency.requirement,
          args.constraints?.[dependency.url]
        )
        results.push(result)
      } catch (error) {
        console.error(`Error checking Swift package ${dependency.url}:`, error)
      }
    }

    return {
      content: [
        {
          type: 'text',
          text: JSON.stringify(results, null, 2),
        },
      ],
    }
  }
}

File: internal/handlers/swift.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"sync"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/sirupsen/logrus"
)

// SwiftHandler handles Swift package version checking
type SwiftHandler struct {
	client HTTPClient
	cache  *sync.Map
	logger *logrus.Logger
}

// NewSwiftHandler creates a new Swift handler
func NewSwiftHandler(logger *logrus.Logger, cache *sync.Map) *SwiftHandler {
	if cache == nil {
		cache = &sync.Map{}
	}
	return &SwiftHandler{
		client: DefaultHTTPClient,
		cache:  cache,
		logger: logger,
	}
}

// GitHubReleaseResponse represents a response from the GitHub API for releases
type GitHubReleaseResponse []struct {
	TagName string `json:"tag_name"`
	Name    string `json:"name"`
	Draft   bool   `json:"draft"`
	Prerelease bool `json:"prerelease"`
	PublishedAt string `json:"published_at"`
}

// GetLatestVersion gets the latest version of Swift packages
func (h *SwiftHandler) GetLatestVersion(ctx context.Context, args map[string]interface{}) (*mcp.CallToolResult, error) {
	h.logger.Info("Getting latest Swift package versions")

	// Parse dependencies
	depsRaw, ok := args["dependencies"]
	if !ok {
		return nil, fmt.Errorf("missing required parameter: dependencies")
	}

	// Convert to []SwiftDependency
	var deps []SwiftDependency
	if depsArr, ok := depsRaw.([]interface{}); ok {
		for _, depRaw := range depsArr {
			if depMap, ok := depRaw.(map[string]interface{}); ok {
				var dep SwiftDependency
				if url, ok := depMap["url"].(string); ok {
					dep.URL = url
				} else {
					continue
				}
				if version, ok := depMap["version"].(string); ok {
					dep.Version = version
				}
				if requirement, ok := depMap["requirement"].(string); ok {
					dep.Requirement = requirement
				}
				deps = append(deps, dep)
			}
		}
	} else {
		return nil, fmt.Errorf("invalid dependencies format: expected array")
	}

	// Parse constraints
	var constraints VersionConstraints
	if constraintsRaw, ok := args["constraints"]; ok {
		if constraintsMap, ok := constraintsRaw.(map[string]interface{}); ok {
			constraints = make(VersionConstraints)
			for name, constraintRaw := range constraintsMap {
				if constraintMap, ok := constraintRaw.(map[string]interface{}); ok {
					var constraint VersionConstraint
					if majorVersion, ok := constraintMap["majorVersion"].(float64); ok {
						majorInt := int(majorVersion)
						constraint.MajorVersion = &majorInt
					}
					if excludePackage, ok := constraintMap["excludePackage"].(bool); ok {
						constraint.ExcludePackage = excludePackage
					}
					constraints[name] = constraint
				}
			}
		}
	}

	// Process each dependency
	results := make([]PackageVersion, 0, len(deps))
	for _, dep := range deps {
		h.logger.WithFields(logrus.Fields{
			"url":     dep.URL,
			"version": dep.Version,
		}).Debug("Processing Swift package")

		// Check if package should be excluded
		if constraint, ok := constraints[dep.URL]; ok && constraint.ExcludePackage {
			results = append(results, PackageVersion{
				Name:       dep.URL,
				Skipped:    true,
				SkipReason: "Package excluded by constraints",
			})
			continue
		}

		// Get latest version
		latestVersion, err := h.getLatestVersion(dep.URL)
		if err != nil {
			h.logger.WithFields(logrus.Fields{
				"url":   dep.URL,
				"error": err.Error(),
			}).Error("Failed to get Swift package info")
			results = append(results, PackageVersion{
				Name:           dep.URL,
				CurrentVersion: StringPtr(dep.Version),
				LatestVersion:  "unknown",
				Registry:       "swift",
				Skipped:        true,
				SkipReason:     fmt.Sprintf("Failed to fetch package info: %v", err),
			})
			continue
		}

		// Apply major version constraint if specified
		if constraint, ok := constraints[dep.URL]; ok && constraint.MajorVersion != nil {
			targetMajor := *constraint.MajorVersion
			latestMajor, _, _, err := ParseVersion(latestVersion)
			if err == nil && latestMajor > targetMajor {
				// Find the latest version with the target major version
				h.logger.WithFields(logrus.Fields{
					"url":          dep.URL,
					"targetMajor":  targetMajor,
					"latestMajor":  latestMajor,
					"latestVersion": latestVersion,
				}).Debug("Applying major version constraint")

				// In a real implementation, this would fetch all versions and filter by major version
				// For now, we'll just append the major version
				latestVersion = fmt.Sprintf("%d.0.0", targetMajor)
			}
		}

		// Add result
		results = append(results, PackageVersion{
			Name:           dep.URL,
			CurrentVersion: StringPtr(dep.Version),
			LatestVersion:  latestVersion,
			Registry:       "swift",
		})
	}

	// Sort results by name
	sort.Slice(results, func(i, j int) bool {
		return strings.ToLower(results[i].Name) < strings.ToLower(results[j].Name)
	})

	return NewToolResultJSON(results)
}

// getLatestVersion gets the latest version of a Swift package
func (h *SwiftHandler) getLatestVersion(packageURL string) (string, error) {
	// Check cache first
	cacheKey := fmt.Sprintf("swift:%s", packageURL)
	if cachedVersion, ok := h.cache.Load(cacheKey); ok {
		h.logger.WithField("url", packageURL).Debug("Using cached Swift package version")
		return cachedVersion.(string), nil
	}

	// Parse GitHub URL
	if !strings.Contains(packageURL, "github.com") {
		return "", fmt.Errorf("only GitHub URLs are supported: %s", packageURL)
	}

	// Extract owner and repo
	parts := strings.Split(packageURL, "/")
	if len(parts) < 5 {
		return "", fmt.Errorf("invalid GitHub URL format: %s", packageURL)
	}

	owner := parts[3]
	repo := parts[4]

	// Construct API URL
	apiURL := fmt.Sprintf("https://api.github.com/repos/%s/%s/releases", owner, repo)
	h.logger.WithFields(logrus.Fields{
		"url":    packageURL,
		"apiURL": apiURL,
	}).Debug("Fetching Swift package releases")

	// Make request
	headers := map[string]string{
		"Accept": "application/vnd.github.v3+json",
	}
	body, err := MakeRequestWithLogger(h.client, h.logger, "GET", apiURL, headers)
	if err != nil {
		return "", fmt.Errorf("failed to fetch Swift package releases: %w", err)
	}

	// Parse response
	var releases GitHubReleaseResponse
	if err := json.Unmarshal(body, &releases); err != nil {
		return "", fmt.Errorf("failed to parse Swift package releases: %w", err)
	}

	// Find latest non-draft, non-prerelease version
	var latestVersion string
	for _, release := range releases {
		if release.Draft || release.Prerelease {
			continue
		}

		version := strings.TrimPrefix(release.TagName, "v")
		if latestVersion == "" {
			latestVersion = version
			continue
		}

		// Compare versions
		result, err := CompareVersions(version, latestVersion)
		if err != nil {
			h.logger.WithFields(logrus.Fields{
				"version1": version,
				"version2": latestVersion,
				"error":    err.Error(),
			}).Debug("Failed to compare versions")
			continue
		}

		if result > 0 {
			latestVersion = version
		}
	}

	if latestVersion == "" {
		// If no releases found, try tags
		tagsURL := fmt.Sprintf("https://api.github.com/repos/%s/%s/tags", owner, repo)
		h.logger.WithFields(logrus.Fields{
			"url":     packageURL,
			"tagsURL": tagsURL,
		}).Debug("Fetching Swift package tags")

		// Make request
		body, err := MakeRequestWithLogger(h.client, h.logger, "GET", tagsURL, headers)
		if err != nil {
			return "", fmt.Errorf("failed to fetch Swift package tags: %w", err)
		}

		// Parse response
		var tags []struct {
			Name string `json:"name"`
		}
		if err := json.Unmarshal(body, &tags); err != nil {
			return "", fmt.Errorf("failed to parse Swift package tags: %w", err)
		}

		// Find latest version
		for _, tag := range tags {
			version := strings.TrimPrefix(tag.Name, "v")
			if latestVersion == "" {
				latestVersion = version
				continue
			}

			// Compare versions
			result, err := CompareVersions(version, latestVersion)
			if err != nil {
				h.logger.WithFields(logrus.Fields{
					"version1": version,
					"version2": latestVersion,
					"error":    err.Error(),
				}).Debug("Failed to compare versions")
				continue
			}

			if result > 0 {
				latestVersion = version
			}
		}
	}

	if latestVersion == "" {
		return "", fmt.Errorf("no releases or tags found for: %s", packageURL)
	}

	// Cache result
	h.cache.Store(cacheKey, latestVersion)

	return latestVersion, nil
}
